ext/dapr-ext-grpc/dapr/ext/grpc/_servicier.py

```diff
@@ -53,19 +68,44 @@ def register_topic(
             pubsub_name: str,
             topic: str,
             cb: TopicSubscribeCallable,
-            metadata: Optional[Dict[str, str]]) -> None:
+            metadata: Optional[Dict[str, str]],
+            rule: Optional[Rule] = None) -> None:
         """Registers topic subscription for pubsub."""
-        pubsub_topic = pubsub_name + DELIMITER + topic
+        topic_key = pubsub_name + DELIMITER + topic
+        pubsub_topic = topic_key + DELIMITER
+        if rule is not None:
+            path = getattr(cb, '__name__', rule.match)
+            pubsub_topic = pubsub_topic + path
         if pubsub_topic in self._topic_map:
-            raise ValueError(f'{topic} is already registered with {pubsub_name}')
+            raise ValueError(
+                f'{topic} is already registered with {pubsub_name}')
         self._topic_map[pubsub_topic] = cb
-        self._registered_topics.append(
-            appcallback_v1.TopicSubscription(
+
+        registered_topic = self._registered_topics_map.get(topic_key)
+        sub: appcallback_v1.TopicSubscription = appcallback_v1.TopicSubscription()
+        rules: List[Tuple[int, appcallback_v1.TopicRule]] = []
+        if not registered_topic:
+            sub = appcallback_v1.TopicSubscription(
                 pubsub_name=pubsub_name,
                 topic=topic,
-                metadata=metadata
+                metadata=metadata,
+                routes=appcallback_v1.TopicRoutes()
             )
-        )
+            registered_topic = _RegisteredSubscription(sub, rules)
+            self._registered_topics_map[topic_key] = registered_topic
+            self._registered_topics.append(sub)
+
+        sub = registered_topic.subscription
+        rules = registered_topic.rules
+
+        if rule:
+            path = getattr(cb, '__name__', rule.match)
+            rules.append((rule.priority, appcallback_v1.TopicRule(
+                match=rule.match, path=path)))
+            rules.sort(key=lambda x: x[0])
+            rs = list(map(lambda r: r[1], rules))
+            del sub.routes.rules[:]
+            sub.routes.rules.extend(rs)
```

Comment:
Should be able to just set rules like this:
```python
sub.routes.rules = rs
```

Was there are reason to empty the list and then add everything back, instead of just replacing the existing list?

Reply:
Apparently, for repeating fields in proto, you cannot directly assign them or you get this error

![image](https://user-images.githubusercontent.com/714039/131411491-d58d03c0-9788-4e2a-a2f6-deaf935957c2.png)

https://stackoverflow.com/questions/23726335/how-to-assign-to-repeated-field

Reply:
Ah, a protobufism, of course.